examples/step-86/step-86.cc

```diff
@@ -0,0 +1,1138 @@
+    // solution vector -- i.e., the variable that stores the current state
+    // of the solution -- is kept inside the time integrator object. This
+    // is useful: If we kept a copy of the solution vector as a member variable
+    // of the current class, one would continuously have to wonder whether it
+    // is still in sync with the version of the solution vector the time
+    // integrator object internally believes is the currently correct
+    // version of this vector. As a consequence, we do not store such a
+    // copy here: Whenever a function requires access to the current value
+    // of the solution, it receives it from the time integrator as an argument.
+    // The same observation can be made about the variable that stores the
+    // current time, or the current length of the time step, or the number
+    // of time steps performed so far: They are all kept inside the time
+    // stepping object.
+    void implicit_function(const double                      time,
+                           const PETScWrappers::MPI::Vector &solution,
+                           const PETScWrappers::MPI::Vector &solution_dot,
+                           PETScWrappers::MPI::Vector       &residual);
+
+    void
+    assemble_implicit_jacobian(const double                      time,
+                               const PETScWrappers::MPI::Vector &solution,
+                               const PETScWrappers::MPI::Vector &solution_dot,
+                               const double                      shift);
+
+    void solve_with_jacobian(const PETScWrappers::MPI::Vector &src,
+                             PETScWrappers::MPI::Vector       &residual);
+
+    PETScWrappers::MPI::SparseMatrix jacobian_matrix;
+
+
+    // In this tutorial program, similar to what we did in step-26, we want to
+    // adapt the mesh at regular time intervals. However, if we are using an
+    // external time stepper, we need to make sure that the time stepper is
+    // aware of the mesh changes -- see the discussion on this topic in the
+    // introduction. In particular, the time stepper must support
+    // the fact that the number of degrees of freedom may change, and it must
+    // support the fact that all internal vectors (e.g., the solution vector and
+    // all intermediate stages used to compute the current time derivative) will
+    // need to be transferred to the new mesh. deal.II does not allow to do this
+    // on each vector separately, so the time stepper needs to be made aware of
+    // when this is happening. The following two functions will then be called
+    // by the time stepper whenever it is time to adapt the mesh (the first
+    // of the two) and as soon as the mesh has been adapted (the second one).
+    // The first receives the current solution as argument; the second receives
+    // a number of vectors that all need to be transfered from the old to the
+    // new mesh.
+    void prepare_for_coarsening_and_refinement(
```

Comment:
> The following two functions...

This is wrong. (including "The first receives ....") The user callback is `decide_for_coarsening_and_refinement` (or however you want to rename it). I think it is misleading not discussing it here, and discuss only  `prepare_for_coarsening_and_refinement` that is called by it below. Can you rearrange the code and make the discussion cleaner?

Reply:
Yes, I think I didn't quite understand the reasons when I wrote this part.